tests/e2e/dynamic_provisioning.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/api/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/driver"
+	"github.com/kubernetes-sigs/aws-ebs-csi-driver/tests/e2e/testsuites"
+)
+
+var _ = Describe("[ebs-csi] Dynamic Provisioning", func() {
+
+	f := framework.NewDefaultFramework("ebs")
+
+	var (
+		cs clientset.Interface
+		ns *v1.Namespace
+	)
+
+	BeforeEach(func() {
+		cs = f.ClientSet
+		ns = f.Namespace
+	})
+
+	for _, t := range driver.ValidVolumeTypes {
+		for _, fs := range driver.ValidFSTypes {
+			volumeType := t
+			fsType := fs
+			Context(fmt.Sprintf("with %q volumeType and %q fsType", volumeType, fsType), func() {
+				It("should create a volume on demand", func() {
+					t := testsuites.StorageClassTest{
+						Client:    cs,
+						Namespace: ns,
+					}
+					TestDynamicProvisioning(t, volumeType, fsType)
+				})
+			})
+		}
+	}
+})
+
+func TestDynamicProvisioning(t testsuites.StorageClassTest, volumeType, fsType string) {
+	storageClass := newStorageClass(t.Namespace.Name, v1.PersistentVolumeReclaimDelete, volumeType, fsType)
+	claimSize := driver.SizeForVolumeType(volumeType)
+	persistentVolumeClaim := newClaim(t.Namespace.Name, storageClass.Name, claimSize)
+
+	storageClass, storageClassCleanup, err := t.ProvisionStorageClass(storageClass)
+	Expect(err).NotTo(HaveOccurred())
+
+	persistedVolumeClaim, persistedVolumeClaimCleanup, err := t.ProvisionPersistentVolumeClaim(persistentVolumeClaim)
+	Expect(err).NotTo(HaveOccurred())
+
+	persistedVolume := t.ValidateProvisionedPersistentVolume(storageClass, persistentVolumeClaim, persistedVolumeClaim)
+
+	podCleanup, err := t.ValidatePodCanWriteAndRead(persistedVolumeClaim)
+	Expect(err).NotTo(HaveOccurred())
+
+	if podCleanup != nil {
+		podCleanup()
+	}
+
+	if persistedVolumeClaimCleanup != nil {
```

Comment:
It would be nicer if these cleanup funcs were used with `defer`  maybe `t.ProvisionPersistentVolumeClaim` can check `PersistentVolumeReclaimPolicy` and return a nil cleanup func if its equal to `v1.PersistentVolumeReclaimDelete`

Reply:
The problem with defer is that the check below, in line 97,  requires persistedVolumeClaimCleanup to have run.